include/deal.II/fe/fe_interface_values.h

```diff
@@ -1662,6 +2357,351 @@ namespace FEInterfaceViews
 
 
 
+  template <int dim, int spacedim>
+  template <class InputVector, class OutputVector>
+  void
+  Scalar<dim, spacedim>::get_local_dof_values(
+    const InputVector &dof_values,
+    OutputVector &     local_dof_values) const
+  {
+    const auto &interface_dof_indices =
+      this->fe_interface->get_interface_dof_indices();
+
+    AssertDimension(interface_dof_indices.size(), local_dof_values.size());
+
+    for (unsigned int i = 0; i < interface_dof_indices.size(); ++i)
+      local_dof_values[i] = dof_values(interface_dof_indices[i]);
+  }
+
+
+
+  template <int dim, int spacedim>
+  template <class InputVector>
+  void
+  Scalar<dim, spacedim>::get_function_interface_values_from_local_dof_values(
+    const bool         here_or_there,
+    const InputVector &local_dof_values,
+    std::vector<solution_value_type<typename InputVector::value_type>> &values)
+    const
+  {
+    const auto &interface_dof_indices =
+      this->fe_interface->get_interface_dof_indices();
+
+    for (unsigned int dof_index = 0; dof_index < interface_dof_indices.size();
+         ++dof_index)
+      for (unsigned int q_index = 0; q_index < values.size(); ++q_index)
+        {
+          values[q_index] += local_dof_values[dof_index] *
+                             value(here_or_there, dof_index, q_index);
+        }
+  }
+
+
+
+  template <int dim, int spacedim>
+  template <class InputVector>
+  void
+  Scalar<dim, spacedim>::get_function_interface_values(
+    const bool         here_or_there,
+    const InputVector &fe_function,
+    std::vector<solution_value_type<typename InputVector::value_type>> &values)
+    const
+  {
+    std::vector<typename InputVector::value_type> local_dof_values(
+      this->fe_interface->n_current_interface_dofs());
+    get_local_dof_values(fe_function, local_dof_values);
+
+    get_function_interface_values_from_local_dof_values(here_or_there,
+                                                        local_dof_values,
+                                                        values);
+  }
+
+
+
+  template <int dim, int spacedim>
+  template <class InputVector>
+  void
+  Scalar<dim, spacedim>::get_jump_in_function_values_from_local_dof_values(
+    const InputVector &local_dof_values,
+    std::vector<solution_value_type<typename InputVector::value_type>> &values)
+    const
+  {
+    const auto &interface_dof_indices =
+      this->fe_interface->get_interface_dof_indices();
+
+    for (unsigned int dof_index = 0; dof_index < interface_dof_indices.size();
+         ++dof_index)
+      for (unsigned int q_index = 0; q_index < values.size(); ++q_index)
+        {
+          values[q_index] +=
+            local_dof_values[dof_index] * jump(dof_index, q_index);
```

Comment:
Hmm... I'm not sure that this is correct. But maybe I'm missing something.

Aren't we looking for the difference between the function value on the one side of the interface and the other? That is, something like
```
jump(u)|_{q} = u_{R} - u_{L} = \sum_{i} U_{i(R)} N_{i(R)} - U_{i(L)} N_{i(L)}
```
per q-point? As it stands, `jump()` gives you the jump in the shape function values, but this jump is applied to both `U_{i(R)}` and `U_{i(L)}` and you end up with something like `(N_{i(R)} - N_{i(L)}) (U_{i(R)} + U_{i(L)}) `. I think that you may need to follow your existing implementation for `jump()` etc., something like
```c++
  Scalar<dim, spacedim>::get_jump_in_function_values_from_local_dof_values(
    const InputVector &local_dof_values,
    std::vector<solution_value_type<typename InputVector::value_type>> &values)
    const
{
    const auto &interface_dof_indices =
      this->fe_interface->get_interface_dof_indices();

    for (unsigned int interface_dof_index = 0; interface_dof_index < interface_dof_indices.size();
         ++interface_dof_index)
   {
    const auto dof_pair = this->fe_interface->dofmap[interface_dof_index];

    for (unsigned int q_index = 0; q_index < values.size(); ++q_index)
   {
    // Function value on one side of the interface
    if (dof_pair[0] != numbers::invalid_unsigned_int)
      value[q_index] +=
        local_dof_values[dof_pair[0]] *
        (*(this->fe_interface->fe_face_values))[extractor].value(dof_pair[0],
                                                                 q_point);

   // Function value on the other side of the interface
    if (dof_pair[1] != numbers::invalid_unsigned_int)
      value[q_index] -=
        local_dof_values[dof_pair[1]] *
        (*(this->fe_interface->fe_face_values_neighbor))[extractor].value(
          dof_pair[1], q_point);
   }
   }
}
```
If this is case, then I believe that the other functions that follow also require correction.

A "simple" way to check this is to have some test with a cell-wise discontinuous function, and manually check this jump and compare it to what is computed with this function.

Reply:
Note: If my above comment is incorrect or not relevant, then the call to jump() and some of the others will be deprecated by #12411. So we'll need to update them once that PR is in.

Reply:
> Hmm... I'm not sure that this is correct. But maybe I'm missing something.
> 
> Aren't we looking for the difference between the function value on the one side of the interface and the other? That is, something like
> 
> ```
> jump(u)|_{q} = u_{R} - u_{L} = \sum_{i} U_{i(R)} N_{i(R)} - U_{i(L)} N_{i(L)}
> ```
> 
> per q-point? As it stands, `jump()` gives you the jump in the shape function values, but this jump is applied to both `U_{i(R)}` and `U_{i(L)}` and you end up with something like `(N_{i(R)} - N_{i(L)}) (U_{i(R)} + U_{i(L)}) `. I think that you may need to follow your existing implementation for `jump()` etc., something like
> 
> ```c++
>   Scalar<dim, spacedim>::get_jump_in_function_values_from_local_dof_values(
>     const InputVector &local_dof_values,
>     std::vector<solution_value_type<typename InputVector::value_type>> &values)
>     const
> {
>     const auto &interface_dof_indices =
>       this->fe_interface->get_interface_dof_indices();
> 
>     for (unsigned int interface_dof_index = 0; interface_dof_index < interface_dof_indices.size();
>          ++interface_dof_index)
>    {
>     const auto dof_pair = this->fe_interface->dofmap[interface_dof_index]; // This is the local DoF index, right?
> 
>     for (unsigned int q_index = 0; q_index < values.size(); ++q_index)
>    {
>     // Function value on one side of the interface
>     if (dof_pair[0] != numbers::invalid_unsigned_int)
>       value[q_index] +=
>         local_dof_values[dof_pair[0]] *
>         (*(this->fe_interface->fe_face_values))[extractor].value(dof_pair[0],
>                                                                  q_point);
> 
>    // Function value on the other side of the interface
>     if (dof_pair[1] != numbers::invalid_unsigned_int)
>       value[q_index] -=
>         local_dof_values[dof_pair[1]] *
>         (*(this->fe_interface->fe_face_values_neighbor))[extractor].value(
>           dof_pair[1], q_point);
>    }
>    }
> }
> ```
> 
> If this is case, then I believe that the other functions that follow also require correction.
> 
> A "simple" way to check this is to have some test with a cell-wise discontinuous function, and manually check this jump and compare it to what is computed with this function.

You are right. Sorry, I should have tested it with DG elements. I will fix it, thanks.

Reply:
It would probably be good to incorporate this sort of question into one of the tests you probably want to add to this PR. A good test would be one where you know the answer -- say, you have two adjacent cells where the values of the solution are piecewise constant and equal to (for example) 1 and 3. Then check that the jump is equal to 2.

Reply:
Hm, @jppelteret , are you sure this is wrong? We have a number of shape functions that live on the union of the two adjacent cells. Then
```
  jump(u_h) = jump(sum_i U_i phi_i) = sum_i u_i jump(phi_i)
```
by linearity. I think that is the formula @zjiaqi2018 implemented, and it looks right to me.

The point is that there isn't a `U_{i,L}` and `U_{i,R}`. We just have nodal values `U_i`, The jump lives only in the shape functions.

Reply:
What one could try to optimize is that the jump for most shape functions is equal either to the value "here", or minus the value "there" (or the other way around), or it is zero (for continuous functions). But that's not always true -- there are funny non-conforming shape functions, and I think that if one were to optimize things, then the `jump()` function would be the place, not this one.

Reply:
What I implemented is essentially the same as what @jppelteret  proposed, because there is always an `invalid_unsigned_int` in the each `dof_pair` for interior faces with DG elements, which means the `jump` function only calculates the jump of one shape function each time. What made me think I was wrong is that I didn't reset `values`, so it just keeps accumulating. I guess adding 
```
if(dof_index == 0)
   values[q_index] = 0.;
```
in the quadrature loop will fix it.

Reply:
You will want to be able to call the `jump()` function also for continuous functions, or non-conforming functions, that exist on both sides. So it isn't true that there is always one `invalid_unsigned_int` in each `dof_pair`.

Reply:
Well, I'm sorry to have opened up a bit of a rabbit hole here and I apologise if my ignorance has led us a bit down the wrong path. But at least its raised a few valuable points, and I can say that it has improved my understanding with respect to how the interface is discretised and what this means when it comes to using "exotic" FE's. @bangerth's comment (https://github.com/dealii/dealii/pull/12603#discussion_r680601418) makes sense now, so if the best solution is actually the original one, then I'd be perfectly fine if we went with that one.